pkg/pool-utils/contracts/controllers/ManagedPoolController.sol

```diff
@@ -0,0 +1,201 @@
+ */
+contract ManagedPoolController is BasePoolController, IControlledManagedPool {
+    using WordCodec for bytes32;
+
+    // There are five managed pool rights: all corresponding to permissioned functions of ManagedPool.
+    struct ManagedPoolRights {
+        bool canChangeWeights;
+        bool canDisableSwaps;
+        bool canSetMustAllowlistLPs;
+        bool canSetCircuitBreakers;
+        bool canChangeTokens;
+    }
+
+    // The minimum weight change duration could be replaced with more sophisticated rate-limiting.
+    uint256 internal immutable _minWeightChangeDuration;
+
+    // Immutable controller state - the first 16 bits are reserved as a bitmap for permission flags
+    // (3 used in the base class; 5 used here), and the remaining 240 bits can be used by derived classes
+    // to store any other immutable data.
+    //
+    //               Managed Pool Controller Permissions             |   Base Controller Permissions  ]
+    // [  240 | 8 bits |  1 bit |  1 bit   | 1 bit | 1 bit |  1 bit  |  1 bit   |  1 bit   |   1 bit  ]
+    // [unused|reserved| tokens | breakers |  LPs  | swaps | weights | metadata | swap fee | transfer ]
+    // |MSB                                                                                        LSB|
+
+    uint256 private constant _CHANGE_WEIGHTS_OFFSET = 3;
+    uint256 private constant _DISABLE_SWAPS_OFFSET = 4;
+    uint256 private constant _MUST_ALLOWLIST_LPS_OFFSET = 5;
+    uint256 private constant _CIRCUIT_BREAKERS_OFFSET = 6;
+    uint256 private constant _CHANGE_TOKENS_OFFSET = 7;
+
+    /**
+     * @dev Pass in the `BasePoolRights` and `ManagedPoolRights` structures, to form the complete set of
+     * immutable rights. Then pass any parameters related to restrictions on those rights. For instance,
+     * a minimum duration if changing weights is enabled.
+     */
+    constructor(
+        BasePoolRights memory baseRights,
+        ManagedPoolRights memory managedRights,
+        uint256 minWeightChangeDuration,
+        address manager
+    ) BasePoolController(encodePermissions(baseRights, managedRights), manager) {
+        _minWeightChangeDuration = minWeightChangeDuration;
+    }
+
+    function encodePermissions(BasePoolRights memory baseRights, ManagedPoolRights memory managedRights)
+        public
+        pure
+        returns (bytes32)
+    {
+        bytes32 permissions = super
+            .encodePermissions(baseRights)
+            .insertBool(managedRights.canChangeWeights, _CHANGE_WEIGHTS_OFFSET)
+            .insertBool(managedRights.canDisableSwaps, _DISABLE_SWAPS_OFFSET)
+            .insertBool(managedRights.canSetMustAllowlistLPs, _MUST_ALLOWLIST_LPS_OFFSET);
+
+        // Last two separated to avoid "stack too deep"
+        return
+            permissions.insertBool(managedRights.canChangeTokens, _CHANGE_TOKENS_OFFSET).insertBool(
+                managedRights.canSetCircuitBreakers,
+                _CIRCUIT_BREAKERS_OFFSET
+            );
```

Comment:
```suggestion
       bytes32 permissions = super.encodePermissions(baseRights);

        return
            permissions
                .insertBool(managedRights.canChangeWeights, _CHANGE_WEIGHTS_OFFSET)
                .insertBool(managedRights.canDisableSwaps, _DISABLE_SWAPS_OFFSET)
                .insertBool(managedRights.canSetMustAllowlistLPs, _MUST_ALLOWLIST_LPS_OFFSET)
                .insertBool(managedRights.canChangeTokens, _CHANGE_TOKENS_OFFSET)
                .insertBool(managedRights.canSetCircuitBreakers, _CIRCUIT_BREAKERS_OFFSET);
```

Reply:
This feels cleaner to me and gets around the stack too deep issues as well.

Reply:
Yes; I agree that's much cleaner. Will do.